src/pudl/analysis/allocate_net_gen.py

```diff
@@ -269,81 +393,58 @@ def allocate_gen_fuel_by_generator_energy_source(
     _ = test_original_gf_vs_the_allocated_by_gens_gf(
         gf=gf, gf_allocated=net_gen_fuel_alloc
     )
-    if drop_interim_cols:
+    # There are a tiny number of records that have NaNs in the prime mover code
+    # and for which the correct prime mover is unclear. Prime mover code is part
+    # of the primary key for this table, so we have to drop them.
+    len_before = net_gen_fuel_alloc.shape[0]
+    net_gen_fuel_alloc = net_gen_fuel_alloc.dropna(subset=["prime_mover_code"])
+    len_after = net_gen_fuel_alloc.shape[0]
+    fraction_dropped = (len_before - len_after) / len_before
+    if fraction_dropped > 5e-5:
+        raise ValueError(
+            "Too many records were found to have a NULL prime_mover_code and "
+            f"dropped. Expected less than 5e-5, but found {fraction_dropped:.1%}."
+        )
```

Comment:
bc generators only ever have one prime mover, we could fill in the prime mover codes instead of dropping these.

Reply:
@cmgosnell I don't feel like I have enough context to know what change to make for this in this PR.

Reply:
There are a handful of records in the allocated dataframe that have NA values for their prime mover codes.  However, we can look up the prime mover code for the generator in that year in the `gens_eia860` dataframe and use that to fill in these few missing values, which means we don't have to lose a little bit of the generation / fuel consumption in the allocation/aggregation process (allowing us to conserve the overall amount of generation and fuel for later validation).

Where's the right place to fill in these missing values?  It seems like it might be earlier in the overall ETL process, like post-harvesting, at which point we have the entity / annual tables available for use in filling in the gen/fuel data table missing PM values.  That way the complete PM values would be available everywhere -- or is this only something that applies after the allocation/aggregation has been done since in the original tables there are no generator IDs to use for looking the PM code up?  In which case doing the PM code filling only after you've gotten the `generator_id` values would make sense.

Reply:
@zaneselvans The `denorm_generation_fuel_combined_eia923` table where the PM code for the allocations comes from doesn't have `generator_id` (it is in `denorm_generation_eia923`). So it makes sense to fill the PM code after you've joined together generation fuel and generation. 

Should I bring in `generators_eia860` in this allocation process to fill in the missing PMs after the generation fuel and generation has been allocated and merged together? Or a different table?

Reply:
for starters, my initial concern doesn't feel like a blocking concern overall. nonetheless this does feel like a good thing to address. if we were to fill in the pm-codes i could see two decent options:
* **at harvesting** we could turn down how strict we are for this variable. right now we have a little function `pudl.transform.eia._manage_strictness` which... manages the strictness 😎  needed for each column we are harvesting. the default is that a record needs to be 70% consistent. There are a few names columns already in here that we turned the strictness of the harvesting to 0% - which is to say that the harvesting process grabs the **most consistent** option found even when it wasn't a plurality of the options found. For the annually-varying/not-widely-reported attributes, this consistency check can be a little weird bc there could be only two or three options in an entity/year. It's possible that turning down the strictness even by a little would result in no-nulls (which we could test somewhere/somehow).
* **some version of bfill in `denorm_generators_eia860`** we have several versions of back-filling/filling in gaps in the output layer mostly in `pudl.output.eia` (ex: `fill_in_missing_ba_codes`, `fill_generator_technology_description`). 

imo the harvesting option would be very easy to implement, with the maybe difficult part being determining where/when to add a "no-nulls in this column" check. the output layer version would be more... like human logic defensible something something if we thought hard about what we think the standard expectation should be. imo a lot of these types of error crop up in years of transition so deciding what is "right" might actually be challenging to do across the board (i.e. whether to backfill or forwardfill). I was initially more inclined to the output option but we could choose wrong so I think turning down the strictness and just getting the most frequently reported value is probably easier and just as likely to produce correct results.

Reply:
(this is all assuming the null pm codes are coming directly from the harvesting process not somewhere else...)